File: training-data/ruby/coin_change.rb
def coin_change(coins, amount)
  dp = Array.new(amount + 1, Float::INFINITY)
  dp[0] = 0
  (1..amount).each do |a|
    coins.each { |coin| dp[a] = [dp[a], dp[a - coin] + 1].min if coin <= a }
  end
  dp[amount] == Float::INFINITY ? -1 : dp[amount]
end

puts coin_change([1, 2, 5], 11)

File: training-data/python/coin_change.py
from typing import List


def coin_change(coins: List[int], amount: int) -> int:
    dp = [0] + [float("inf")] * amount
    for a in range(1, amount + 1):
        for coin in coins:
            if coin <= a:
                dp[a] = min(dp[a], dp[a - coin] + 1)
    return -1 if dp[amount] == float("inf") else dp[amount]


if __name__ == "__main__":
    print(coin_change([1, 2, 5], 11))
